hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/schema/MemberSchemaService.java

```diff
@@ -131,47 +95,194 @@ public void put(Schema schema) {
+                .collect(Collectors.toList());
+
+        persistAllSchemasToHotRestart(schemas);
+        replicator.setReplications(replications);
+    }
+
+    /**
+     * Called on the participant members to prepare for the replication process
+     * by:
+     * <ul>
+     *     <li>Putting the schema to an in-memory registry</li>
+     *     <li>Persisting the schema to HotRestart(if available)</li>
+     *     <li>Marking the schema as {@link SchemaReplicationStatus#PREPARED}
+     *     </li>
+     * </ul>
+     *
+     * @param schema to be prepared for replication.
+     */
+    public void onSchemaPreparationRequest(Schema schema) {
+        // If it is already PREPARED or REPLICATED, do nothing
+        if (replicator.getReplicationStatus(schema) != null) {
+            return;
+        }
+
+        putLocal(schema);
+        persistSchemaToHotRestart(schema);
+        replicator.markSchemaAsPrepared(schema);
+    }
+
+    /**
+     * Called on the participant members to mark the replication status of the
+     * schema as {@link SchemaReplicationStatus#REPLICATED}.
+     *
+     * @param schemaId                  of the schema that is replicated to the
+     *                                  cluster.
+     * @param isLocalOperationExecution {@code true} if the call is made through
+     *                                  local execution of the operation
+     */
+    public void onSchemaAckRequest(long schemaId, boolean isLocalOperationExecution) {
+        replicator.markSchemaAsReplicated(schemaId, isLocalOperationExecution);
+    }
+
+    /**
+     * Called when the schemas are read from the HotRestart data.
+     * <p>
+     * Each schema will be put into the in-memory registry and will be marked as
+     * {@link SchemaReplicationStatus#PREPARED}, as we cannot distinguish
+     * schemas that are {@link SchemaReplicationStatus#PREPARED} and
+     * {@link SchemaReplicationStatus#REPLICATED} from the HotRestart data.
+     *
+     * @param schemas read from the HotRestart data
+     */
+    public void onHotRestartRestore(Collection<Schema> schemas) {
+        for (Schema schema : schemas) {
+            putLocal(schema);
+            replicator.markSchemaAsPrepared(schema);
+        }
+    }
+
+    @Override
+    public Runnable prepareMergeRunnable() {
+        // Called in the member of the smaller cluster that will join the
+        // larger cluster.
+
+        // Since this class implements CoreService interface, we are sure that
+        // this task will run before any task that would merge the data
+        // (like MapService etc.) from smaller cluster to larger cluster.
+        // That makes sure that the schema is replicated before the data.
+
+        // List of schemas that are replicated in the smaller cluster.
+        Collection<SchemaReplication> replications = replicator.getReplications();
+
+        return new SchemaReplicationsMerger(replicator, replications);
```

Comment:
I think that you should pass the replications of the larger cluster but you are passing the replications of smaller cluster here, right?

Reply:
Replications of the larger cluster are passed to the joining member as a pre-join operation. This is for the schemas of the smaller cluster as you noted